src/components/games/rps.ts

```diff
@@ -0,0 +1,333 @@
+  User,
+} from 'discord.js';
+import { getCoinEmoji, getEmojiByName } from '../emojis';
+import { SapphireMessageResponse, SapphireSentMessageType } from '../../codeyCommand';
+import { adjustCoinBalanceByUserId, UserCoinEvent } from '../coin';
+
+class RpsGameTracker {
+  // Key = id, Value = game
+  games: Map<number, RpsGame>;
+
+  constructor() {
+    this.games = new Map<number, RpsGame>();
+  }
+
+  getGameFromId(id: number): RpsGame | undefined {
+    return this.games.get(id);
+  }
+
+  runFuncOnGame(gameId: number, func: (game: RpsGame) => void): void {
+    func(this.getGameFromId(gameId)!);
+  }
+
+  /*
+    Starts a RPS game
+  */
+  async startGame(
+    bet: number,
+    channelId: string,
+    player1User: User,
+    player2User?: User,
+  ): Promise<RpsGame> {
+    const db = await openDB();
+    const result = await db.run(
+      `
+    INSERT INTO rps_game_info (player1_id, player2_id, bet)
+    VALUES (?, ?, ?)
+  `,
+      [player1User.id, player2User?.id, bet],
+    );
+    // get last inserted ID
+    const id = result.lastID;
+    if (id) {
+      const state: RpsGameState = {
+        player1Id: player1User.id,
+        player1Username: player1User.username,
+        player2Id: player2User?.id,
+        player2Username: player2User?.username ?? `Codey ${getEmojiByName('codeyLove')}`,
+        bet,
+        status: 0,
+        player1Sign: 0,
+        player2Sign: 0,
+      };
+      const game = new RpsGame(id!, channelId, state);
+      this.games.set(id, game);
+      return game;
+    }
+    throw new Error('Something went wrong when starting the RPS game');
+  }
+
+  async endGame(gameId: number): Promise<void> {
+    const db = await openDB();
+    const game = this.getGameFromId(gameId);
+    // Don't do anything if game status is still pending
+    if (game?.state.status === RpsGameStatus.Pending) return;
+    // Update winnings
+    switch (game?.state.status) {
+      case RpsGameStatus.Player1Win:
+        await adjustCoinBalanceByUserId(game.state.player1Id, game.state.bet, UserCoinEvent.RpsWin);
+        if (game.state.player2Id) {
+          await adjustCoinBalanceByUserId(
+            game.state.player2Id,
+            -game.state.bet,
+            UserCoinEvent.RpsLoss,
+          );
+        }
+        break;
+      case RpsGameStatus.Player2Win:
+        await adjustCoinBalanceByUserId(
+          game.state.player1Id,
+          -game.state.bet,
+          UserCoinEvent.RpsLoss,
+        );
+        if (game.state.player2Id) {
+          await adjustCoinBalanceByUserId(
+            game.state.player2Id,
+            game.state.bet,
+            UserCoinEvent.RpsWin,
+          );
+        }
+        break;
+      case RpsGameStatus.Draw:
+        if (!game.state.player2Id) {
+          await adjustCoinBalanceByUserId(
+            game.state.player1Id,
+            -game.state.bet / 2,
+            UserCoinEvent.RpsDrawAgainstCodey,
+          );
+        }
+        break;
+      default:
```

Comment:
I'll remove default, but for consistency I think I'll keep the `break;`.

Reply:
Actually, default is still needed because otherwise if a case that does not appear in the listed options arises, the switch statement would throw an error.

Reply:
That's odd; I made a [small program](https://www.typescriptlang.org/play?#code/MYewdgzgLgBANgUylBAnAXDaqCWYDmMAvDAOQBepA3AFA0QDuOUwAFjABSLJoCUMAbxowRMYAEMICMuNLphoxaEghEAOjgh8HUrN61FigEaoE4gNYHRAemswAJggBm4gK5wo8gL51lEVQgaWjpGpPpAA) to test it and it doesn't seem to throw an error?

Reply:
Weird, I suppose TS behaves differently from other languages, like C# (which does throw an error).